utils/global-state-update-gen/src/generic.rs

```diff
@@ -172,22 +172,42 @@ fn gen_snapshot_from_old(
         .collect();
 
     for recipients in snapshot.values_mut() {
-        recipients.retain(|public_key, recipient| match stakes_map.get(public_key) {
-            Some(validator) if validator.bonded_amount == U512::zero() => false,
-            Some(validator) => {
-                *recipient = SeigniorageRecipient::new(
-                    validator.bonded_amount,
-                    validator
-                        .delegation_rate
-                        .unwrap_or(*recipient.delegation_rate()),
-                    validator
-                        .delegators_map()
-                        .unwrap_or_else(|| recipient.delegator_stake().clone()),
+        recipients.retain(
+            |public_key, recipient| match validators_map.get(public_key) {
+                Some(validator) if validator.bonded_amount == U512::zero() => false,
+                Some(validator) => {
+                    *recipient = SeigniorageRecipient::new(
+                        validator.bonded_amount,
+                        validator
+                            .delegation_rate
+                            .unwrap_or(*recipient.delegation_rate()),
+                        validator
+                            .delegators_map()
+                            .unwrap_or_else(|| recipient.delegator_stake().clone()),
+                    );
+                    true
+                }
+                None => true,
+            },
+        );
+
+        // add the validators that weren't present in the old snapshot
+        for (public_key, validator) in &validators_map {
+            if recipients.contains_key(public_key) {
+                continue;
+            }
+
+            if validator.bonded_amount != U512::zero() {
+                recipients.insert(
+                    public_key.clone(),
+                    SeigniorageRecipient::new(
+                        validator.bonded_amount,
+                        validator.delegation_rate.unwrap_or_default(),
+                        validator.delegators_map().unwrap_or_default(),
+                    ),
                 );
-                true
             }
-            None => true,
-        });
+        }
```

Comment:
Is there a function that would "zip" two collections over the set of keys and give you `Left(Option<value>)` and `Right(Option<value>)` ? That way you wouldn't have to iterate twice and it would be easier to understand.

Reply:
I don't know, but it's a good idea.